docs/type-lists.md

```diff
@@ -0,0 +1,75 @@
+# Type Lists
+
+`T...` means a type list called `T...`. In a generic, if it is given a constraint, the constraint applies to each type in the list, not the list itself. So `T...: U` means a type list of any length such that each element is a subtype of `U`. The type `()` is an empty type list. Unlike a singleton type, this is not a subtype of the empty interface.
+
+`A...` is a tuple of zero or more types. `(A, B...)` is a tuple of one or more types. That is, the type list is flattened into the tuple. `A...->B` is a function type with zero or more arguments. `(A, B...)->C` is a function type with one or more arguments.
+
+`A...-->B` is a curried function type (note the use of `-->` instead of `->`). That is, the type list is flattened into the arrow. If `A...` is bound to `(C, D)`, you would get `C->D->B`.
+
+Function parameters and results may have type annotations containing type lists. Calling a function does a destructuring bind of the tuple argument. Using the result also does a destructuring bind.
+
+Tuple flattening also happens at the value level. This flattening is type-directed: given `x: A...`, uses of `x` are expanded to a sequence of values within the enclosing tuple. For example, given `x: A, y: B...`, if `B...` is bound to a tuple of length 2, `(x, y)` is a tuple of length 3. If instead `B...` is bound to a tuple of length 1, then `(x, y)` and `(x, (y))` are identical.
```

Comment:
Perhaps also add to the "given `x: A, y: B...`, if `B...` is bound to a tuple of length 2," case that `(x, (y))` is not `(x, y)` but rather a pair whose second component is a pair?

Reply:
Good call.